### PR TITLE
Cancel requests that are queued for a client/handler on error

### LINKED ISSUE #2478: [BUG] When tasks complete, cancel pending UCX requests
Splitting off a piece of pending requests when tasks complete (from https://github.com/NVIDIA/spark-rapids/issues/1178):

> The lowest hanging fruit are client request that are waiting for bounce buffers: No resources have been acquired, and also no risk of a server having pending sends. So I think this part is doable today. This could cover a very large number of messages as well. Just thinking in terms of bounce buffer lengths.. we could have several bounce buffer lengths not ready to be received in the queue.

This issue is to track just this piece.

### PR DESCRIPTION
**Pull request by abellina**
Closes #2478 

This handles a performance bug that will cancel queued requests if a peer leaves. These are not requests in flight, so a `recv` has not been issued.

### CODE REVIEW

File: shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala [hunk 1]
```diff
@@ -88,6 +88,22 @@ class UCXShuffleTransport(shuffleServerId: BlockManagerId, rapidsConf: RapidsCon
...
+    1000,
+    (t: PendingTransferRequest, t1: PendingTransferRequest) => {
+      if (t.getLength < t1.getLength) {
+        -1;
+      } else if (t.getLength > t1.getLength) {
+        1;
+      } else {
+        0
+      }
+    })
```
Comment (jlowe) [inline shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala:101]:
```suggestion
    (t: PendingTransferRequest, t1: PendingTransferRequest) =>
      java.lang.Long.compare(t.getLength, t1.getLength))
```

File: shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala [hunk 2]
```diff
@@ -469,11 +479,39 @@ class UCXShuffleTransport(shuffleServerId: BlockManagerId, rapidsConf: RapidsCon
...
       altList.notifyAll()
     }
 
+  override def cancelPending(handler: RapidsShuffleFetchHandler): Unit = {
+    altList.synchronized {
+      if (validHandlers.contains(handler)) {
+        // This is expensive, but will be refactored with a queue per client.
+        // As it stands, in the good case it should be invoked once per task/peer,
+        // on task completion, and `altList` should be empty
+        // will be skipped.
```
Comment (jlowe) [inline shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala:494]: I had a hard time following this sentence -- "should be empty will be skipped."

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 3]
```diff
@@ -258,16 +258,20 @@ class RapidsShuffleIterator(
...
+
+            // tell the client to cancel pending requests
+            client.cancelPending(this)
+          }
         }
 
         logInfo(s"Client $blockManagerId triggered, for ${shuffleRequestsMapIndex.size} blocks")
         client.doFetch(shuffleRequestsMapIndex.map(_.id), handler)
-        clients = clients :+ client
+        clientAndHandlers = clientAndHandlers :+ ((client, handler))
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:269]:
```suggestion
        clientAndHandlers = clientAndHandlers :+ (client, handler)
```

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:269]: In this case, `clientAndHandlers` is a `Seq` of tuple, so the original signature is required according to my compiler.

**Comment by abellina** [comment]
I went ahead and applied your other changes in: https://github.com/NVIDIA/spark-rapids/pull/2553/commits/161c6faa09e0ee36243e0a4217071d5ce05fb9a3 @jlowe.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala (hunk 3 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:269]: Weird, it works for me in the Spark Scala REPL:
```
scala> var x: Seq[(Int, Int)] = Nil
x: Seq[(Int, Int)] = List()

scala> x = x :+ (1, 2)
x: Seq[(Int, Int)] = List((1,2))

scala> x = x :+ ((1, 2))
x: Seq[(Int, Int)] = List((1,2), (1,2))

scala> x = x :+ (1, 2)
x: Seq[(Int, Int)] = List((1,2), (1,2), (1,2))
```

The `:+` operator adds a single object to the sequence, and what's being added is a tuple.  Oh well no worries, if it's necessary it's necessary.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala (hunk 3 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:269]: I can reproduce your case in the scala repl. But when I try with the `Seq` in the iterator, this is what I get:

```
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala
index 969f2c955..8178ef664 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala
@@ -266,7 +266,7 @@ class RapidsShuffleIterator(
 
         logInfo(s"Client $blockManagerId triggered, for ${shuffleRequestsMapIndex.size} blocks")
         client.doFetch(shuffleRequestsMapIndex.map(_.id), handler)
-        clientAndHandlers = clientAndHandlers :+ ((client, handler))
+        clientAndHandlers = clientAndHandlers :+ (client, handler)
       }
     }
```

```
[ERROR] [Error] /home/abellina/work/rapids-plugin-4-spark-latest/sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:269: too many arguments (2) for method :+: (elem: (com.nvidia.spark.rapids.shuffle.RapidsShuffleClient, com.nvidia.spark.rapids.shuffle.RapidsShuffleFetchHandler))(implicit bf: scala.collection.generic.CanBuildFrom[Seq[(com.nvidia.spark.rapids.shuffle.RapidsShuffleClient, com.nvidia.spark.rapids.shuffle.RapidsShuffleFetchHandler)],(com.nvidia.spark.rapids.shuffle.RapidsShuffleClient, com.nvidia.spark.rapids.shuffle.RapidsShuffleFetchHandler),That])That
```

**Pull request merged**